pkg/pillar/cmd/downloader/download.go

```diff
@@ -19,41 +24,91 @@ import (
 	"github.com/lf-edge/eve-libs/zedUpload/types"
 	"github.com/lf-edge/eve/pkg/pillar/controllerconn"
 	"github.com/lf-edge/eve/pkg/pillar/netdump"
+	utils "github.com/lf-edge/eve/pkg/pillar/utils/file"
 )
 
 const AWSS3IDENTIFIER = "amazonaws"
 
+// on-disk format with self-check
+type progressFile struct {
+	Version int                   `json:"version"`
+	Hash    string                `json:"hash"` // sha256 of Parts JSON encoded as hex
+	Parts   types.DownloadedParts `json:"parts"`
+}
+
+func hashParts(p types.DownloadedParts) string {
+	b, err := json.Marshal(p)
+	if err != nil {
+		log.Error("Failed to marshal downloaded parts")
+		return ""
+	}
+	sum := sha256.Sum256(b)
+	return hex.EncodeToString(sum[:])
+}
+
+// helper to read+verify one file
+func read(p string) (types.DownloadedParts, bool) {
+	var zero types.DownloadedParts
+	fd, err := os.Open(p)
+	if err != nil {
+		return zero, false
+	}
+	defer fd.Close()
+	data, err := io.ReadAll(fd)
+	if err != nil {
+		return zero, false
+	}
+	var pf progressFile
+	if err := json.Unmarshal(data, &pf); err != nil {
+		return zero, false
+	}
+	if pf.Hash != hashParts(pf.Parts) {
+		log.Warnf("progress file hash mismatch for %s (discarding)", p)
+		return zero, false
+	}
+	return pf.Parts, true
+}
+
 func loadDownloadedParts(locFilename string) types.DownloadedParts {
-	var downloadedParts types.DownloadedParts
-	fd, err := os.Open(locFilename + progressFileSuffix)
-	if err == nil {
-		decoder := json.NewDecoder(fd)
-		err = decoder.Decode(&downloadedParts)
-		if err != nil {
-			log.Errorf("failed to decode progress file: %s", err)
-		}
-		err := fd.Close()
-		if err != nil {
-			log.Errorf("failed to close progress file: %s", err)
+	var zero types.DownloadedParts
+	path := locFilename + progressFileSuffix
+	tmp := path + tmpFileSuffix
+
+	// try the main file
+	if parts, ok := read(path); ok {
+		return parts
+	}
+
+	// recover from a valid, newer tmp (leftover from crash between fsync and rename)
+	dir := filepath.Dir(path)
+	stMain, errMain := os.Stat(path)
+	stTmp, errTmp := os.Stat(tmp)
+	// main file does not exist or tmp exists and is newer than main -> promote
+	if (errMain != nil && errors.Is(errMain, fs.ErrNotExist)) ||
```

Comment:
are you sure you have to check for `nil`?

```suggestion
	if errors.Is(errMain, fs.ErrNotExist) ||
```

Reply:
sorry I missed this comment, added my own https://github.com/lf-edge/eve/pull/5254#discussion_r2393867586, but I think you should check for `nil`.